### AINAHACK - 2024_accions_formatives_previstes_consorci.csv

### IMPORTAR PAQUETES

In [183]:
import numpy as np
import pandas as pd
from janitor import clean_names

# Automcompletar rápido
%config IPCompleter.greedy=True

### IMPORTAR LOS DATOS

Sustituir el nombre_fichero_datos.

In [184]:
nombre_fichero_datos = '2024_accions_formatives_previstes_consorci.csv'

Cargar los datos.

In [185]:
# Fix the path to use proper directory separators and a more robust path construction
import os

# Update the project path with the complete path
ruta_proyecto = os.path.join(os.path.expanduser("~"), "Downloads", "YTT", "AINAHACK", "AinaHack-Data")

# Update the path to match the actual file location
ruta_completa = os.path.join(ruta_proyecto, "02_Datos", "01_Originales", nombre_fichero_datos)

# Add error handling for file reading
try:
    df = pd.read_csv(ruta_completa, encoding='utf-8')
    print(f"Archivo cargado exitosamente desde: {ruta_completa}")
except FileNotFoundError:
    print(f"Error: No se encontró el archivo en la ruta: {ruta_completa}")
    print("Por favor, verifica que:")
    print("1. La ruta del proyecto es correcta:", ruta_proyecto)
    print("2. El archivo existe en la carpeta 02_Datos/01_Originales")
    print("3. El nombre del archivo es exactamente '2024_accions_formatives_previstes_consorci.csv'")

df = pd.read_csv(ruta_completa,index_col=0)
df

Archivo cargado exitosamente desde: C:\Users\Usuario\Downloads\YTT\AINAHACK\AinaHack-Data\02_Datos\01_Originales\2024_accions_formatives_previstes_consorci.csv


FAMÍLIA PROFESSIONAL  \
IDENTIFICADOR                                            
258262         Serveis socioculturals i a la comunitat   
258172         Serveis socioculturals i a la comunitat   
251433                            Hostaleria i turisme   
251440                            Hostaleria i turisme   
256855                          Administració i gestió   
...                                                ...   
242555                                         Sanitat   
256225                                         Agrària   
256240                         Edificació i obra civil   
256241                        Seguretat i medi ambient   
256239                         Edificació i obra civil   

                                   ÀREA PROFESSIONAL  \
IDENTIFICADOR                                          
258262                           Formació i educació   
258172                           Formació i educació   
251433                                   Restauració   
251440                                   Restauració   
256855         Gestió de la informació i comunicació   
...                                              ...   
242555                             Atenció sanitària   
256225                                     Ramaderia   
256240                     Maquinària de construcció   
256241                         Seguretat i prevenció   
256239                     Maquinària de construcció   

                                                  NOM DE L'ACCIÓ  \
IDENTIFICADOR                                                      
258262         SSCE0110_MP0353_Mòdul de pràctiques profession...   
258172         SSCE0110_MP0353_Mòdul de pràctiques profession...   
251433                    Pastisseria: Postres per a restauració   
251440                       Ofertes gastronòmiques de formatges   
256855         ADGG0208_UF0323_Aplicacions informàtiques per ...   
...                                                          ...   
242555            Cures en pacients amb risc de lesions cutànies   
256225         Benestar animal: especialitat en explotacions ...   
256240         Operador/a de plataformes mòbils de personal (...   
256241               Seguretat i tècniques en treballs en alçada   
256239         Actualització plataformes elevadores mòbils de...   

              NIF_ENTITAT_FORMACIÓ  \
IDENTIFICADOR                        
258262                   P7590004C   
258172                   G60374022   
251433                   A08311342   
251440                   A08311342   
256855                   B66117797   
...                            ...   
242555                   G58825811   
256225                   B62278783   
256240                   B62278783   
256241                   B62278783   
256239                   B62278783   

                                                      ENTITAT  CODI_MUNICIPI  \
IDENTIFICADOR                                                                  
258262           Institut Municipal d'Ocupació Salvador Segui          25120   
258172                                                 JOVIAT           8113   
251433                                              GRUP CETT           8019   
251440                                              GRUP CETT           8019   
256855         P.H.R.O. TRAINING CONSULTANTS & PARTNERS, S.L.           8266   
...                                                       ...            ...   
242555                                                   ACRA           8019   
256225                                               CRITERIA           8019   
256240                                               CRITERIA           8019   
256241                                               CRITERIA           8019   
256239                                               CRITERIA           8019   

                        NOM_MUNICIPI  CODI_COMARCA        NOM_COMARCA  \
IDENTIFICADOR                                           

### CORRECCIÓN DE NOMBRES DE VARIABLES

**Corrección automática**

In [186]:
df = clean_names(df) # Limpieza de nombres de columnas de forma automática
df

familia_professional  \
IDENTIFICADOR                                            
258262         Serveis socioculturals i a la comunitat   
258172         Serveis socioculturals i a la comunitat   
251433                            Hostaleria i turisme   
251440                            Hostaleria i turisme   
256855                          Administració i gestió   
...                                                ...   
242555                                         Sanitat   
256225                                         Agrària   
256240                         Edificació i obra civil   
256241                        Seguretat i medi ambient   
256239                         Edificació i obra civil   

                                   area_professional  \
IDENTIFICADOR                                          
258262                           Formació i educació   
258172                           Formació i educació   
251433                                   Restauració   
251440                                   Restauració   
256855         Gestió de la informació i comunicació   
...                                              ...   
242555                             Atenció sanitària   
256225                                     Ramaderia   
256240                     Maquinària de construcció   
256241                         Seguretat i prevenció   
256239                     Maquinària de construcció   

                                                   nom_de_laccio  \
IDENTIFICADOR                                                      
258262         SSCE0110_MP0353_Mòdul de pràctiques profession...   
258172         SSCE0110_MP0353_Mòdul de pràctiques profession...   
251433                    Pastisseria: Postres per a restauració   
251440                       Ofertes gastronòmiques de formatges   
256855         ADGG0208_UF0323_Aplicacions informàtiques per ...   
...                                                          ...   
242555            Cures en pacients amb risc de lesions cutànies   
256225         Benestar animal: especialitat en explotacions ...   
256240         Operador/a de plataformes mòbils de personal (...   
256241               Seguretat i tècniques en treballs en alçada   
256239         Actualització plataformes elevadores mòbils de...   

              nif_entitat_formacio  \
IDENTIFICADOR                        
258262                   P7590004C   
258172                   G60374022   
251433                   A08311342   
251440                   A08311342   
256855                   B66117797   
...                            ...   
242555                   G58825811   
256225                   B62278783   
256240                   B62278783   
256241                   B62278783   
256239                   B62278783   

                                                      entitat  codi_municipi  \
IDENTIFICADOR                                                                  
258262           Institut Municipal d'Ocupació Salvador Segui          25120   
258172                                                 JOVIAT           8113   
251433                                              GRUP CETT           8019   
251440                                              GRUP CETT           8019   
256855         P.H.R.O. TRAINING CONSULTANTS & PARTNERS, S.L.           8266   
...                                                       ...            ...   
242555                                                   ACRA           8019   
256225                                               CRITERIA           8019   
256240                                               CRITERIA           8019   
256241                                               CRITERIA           8019   
256239                                               CRITERIA           8019   

                        nom_municipi  codi_comarca        nom_comarca  \
IDENTIFICADOR                                           

**Corrección manual**

In [187]:
# df.rename(columns = {'fomacion':'formacion',
#                      'prestamo_hipotecario':'hipoteca',
#                      'prestamo_personal':'consumo'}, inplace=True)
# df

### VISIÓN GENERAL

In [188]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4952 entries, 258262 to 256239
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   familia_professional  4952 non-null   object 
 1   area_professional     4952 non-null   object 
 2   nom_de_laccio         4952 non-null   object 
 3   nif_entitat_formacio  4952 non-null   object 
 4   entitat               4952 non-null   object 
 5   codi_municipi         4952 non-null   int64  
 6   nom_municipi          4952 non-null   object 
 7   codi_comarca          4952 non-null   int64  
 8   nom_comarca           4952 non-null   object 
 9   modalitat             4952 non-null   object 
 10  tipus_de_formacio     3055 non-null   object 
 11  hores                 4952 non-null   int64  
 12  web                   4938 non-null   object 
 13  correu_electronic     4938 non-null   object 
 14  telefon               4938 non-null   float64
dtypes: float64(1), int6

### TIPOS DE DATOS

- Identificación

In [189]:
df.dtypes

familia_professional     object
area_professional        object
nom_de_laccio            object
nif_entitat_formacio     object
entitat                  object
codi_municipi             int64
nom_municipi             object
codi_comarca              int64
nom_comarca              object
modalitat                object
tipus_de_formacio        object
hores                     int64
web                      object
correu_electronic        object
telefon                 float64
dtype: object

- Corrección

In [190]:
# Convertir tipos de datos más apropiados
df = df.astype({
    # Mantener como object (campos de texto)
    'familia_professional': 'object',    # Categoría de familia profesional
    'area_professional': 'object',       # Área profesional
    'nom_de_laccio': 'object',          # Nombre de la acción
    'nif_entitat_formacio': 'object',   # NIF (mantener como texto para preservar formato)
    'entitat': 'object',                # Nombre de la entidad
    'nom_municipi': 'object',           # Nombre del municipio
    'nom_comarca': 'object',            # Nombre de la comarca
    'modalitat': 'object',              # Modalidad
    'tipus_de_formacio': 'object',      # Tipo de formación
    'web': 'object',                    # URL web
    'correu_electronic': 'object',      # Email
    
    # Mantener como int64 (códigos y valores enteros)
    'codi_municipi': 'int64',           # Código de municipio
    'codi_comarca': 'int64',            # Código de comarca
    'hores': 'int64',                   # Horas de formación
    
    # Convertir teléfono a string/object
    'telefon': 'object'                 # Teléfono (mejor como texto para preservar formato)
})

df.dtypes

familia_professional    object
area_professional       object
nom_de_laccio           object
nif_entitat_formacio    object
entitat                 object
codi_municipi            int64
nom_municipi            object
codi_comarca             int64
nom_comarca             object
modalitat               object
tipus_de_formacio       object
hores                    int64
web                     object
correu_electronic       object
telefon                 object
dtype: object

- tipus_de_formacio tiene 1897 valores nulos (3055 no nulos de 4952)
- web, correu_electronic y telefon tienen 14 valores nulos cada uno (4938 no nulos de 4952)
- El resto de campos están completos (4952 no nulos)

El cambio más significativo sugerido es convertir telefon de float64 a object porque:
- Los números de teléfono no son realmente números para operar
- Pueden tener formatos específicos (prefijos, guiones, etc.)
- Pueden empezar por 0, lo cual se perdería en formato numérico

### VALORES ÚNICOS

- Identificación

In [191]:
df.nunique().sort_values()

modalitat                  3
tipus_de_formacio          4
codi_comarca              21
nom_comarca               22
familia_professional      24
codi_municipi             38
nom_municipi              39
hores                     61
area_professional         82
telefon                   99
correu_electronic        101
web                      101
nif_entitat_formacio     104
entitat                  104
nom_de_laccio           1344
dtype: int64

In [192]:
columnas_un_valor = [col for col in df.columns if df[col].nunique() == 1]
# ambit, tipusequipament, tipusgeneral, titularitat
print("Valores únicos en columnas con un solo valor:")
for col in columnas_un_valor:
    print(f"{col}: {df[col].unique()[0]}")

Valores únicos en columnas con un solo valor:


In [193]:
# Variables con pocas categorías
variables_categoricas = {
    'modalitat': 3,           # Solo 3 tipos de modalidades
    'tipus_de_formacio': 4    # 4 tipos de formación
}

# Ver las categorías específicas
for var in variables_categoricas:
    print(f"\nCategorías en {var}:")
    print(df[var].value_counts())


Categorías en modalitat:
modalitat
Presencial      3821
Teleformació    1116
Mixta             15
Name: count, dtype: int64

Categorías en tipus_de_formacio:
tipus_de_formacio
Altra Formació Professionalitzadora                                                                                                          2223
Certificats de Professionalitat                                                                                                               530
Acreditacions oficials que capaciten per a un lloc de treball, formació regulada per conveni col·lectiu o certificacions de fabricant TIC     172
Formació d'idiomes preparatòria per certificacions segons Marc Comú Europeu de Referència de les Llengües                                     130
Name: count, dtype: int64


In [194]:
# 1. Imputar modalitat por la moda
moda_modalitat = df['modalitat'].mode()[0]  # Obtiene la moda ('Presencial')
df['modalitat'] = df['modalitat'].fillna(moda_modalitat)

print("Verificación después de imputar modalitat:")
print(df['modalitat'].value_counts())
print(f"Nulos en modalitat: {df['modalitat'].isna().sum()}\n")

# 2. Análisis profundo de tipus_de_formacio
print("Análisis de tipus_de_formacio:")
print("\na) Distribución actual:")
print(df['tipus_de_formacio'].value_counts())
print(f"\nb) Valores nulos: {df['tipus_de_formacio'].isna().sum()} ({(df['tipus_de_formacio'].isna().sum()/len(df)*100):.1f}%)")

# 3. Análisis cruzado
print("\nc) Relación entre tipus_de_formacio y modalitat:")
print(pd.crosstab(df['tipus_de_formacio'], df['modalitat'], margins=True))

# 4. Análisis de patrones en los nulos
print("\nd) Características de registros con tipus_de_formacio nulo:")
nulos_tipus = df[df['tipus_de_formacio'].isna()]
print("\nDistribución de modalitat en registros con tipus_de_formacio nulo:")
print(nulos_tipus['modalitat'].value_counts())
print("\nDistribución de familia_professional en registros con tipus_de_formacio nulo (top 5):")
print(nulos_tipus['familia_professional'].value_counts().head())

Verificación después de imputar modalitat:
modalitat
Presencial      3821
Teleformació    1116
Mixta             15
Name: count, dtype: int64
Nulos en modalitat: 0

Análisis de tipus_de_formacio:

a) Distribución actual:
tipus_de_formacio
Altra Formació Professionalitzadora                                                                                                          2223
Certificats de Professionalitat                                                                                                               530
Acreditacions oficials que capaciten per a un lloc de treball, formació regulada per conveni col·lectiu o certificacions de fabricant TIC     172
Formació d'idiomes preparatòria per certificacions segons Marc Comú Europeu de Referència de les Llengües                                     130
Name: count, dtype: int64

b) Valores nulos: 1897 (38.3%)

c) Relación entre tipus_de_formacio y modalitat:
modalitat                                           Mixta  Presencial

- Anàlisis de la variable tipus_formacio

In [195]:
# Ver los registros con tipus_de_formacio nulo
nulos_analysis = df[df['tipus_de_formacio'].isna()]

print("Análisis de registros con tipus_de_formacio nulo:")
print("\n1. Distribución por familia_professional:")
print(nulos_analysis['familia_professional'].value_counts())

print("\n2. Distribución por area_professional:")
print(nulos_analysis['area_professional'].value_counts().head(10))  # Top 10

print("\n3. Ejemplo de algunos registros completos:")
print(nulos_analysis[['familia_professional', 'area_professional', 'nom_de_laccio', 'modalitat']].head())

Análisis de registros con tipus_de_formacio nulo:

1. Distribución por familia_professional:
familia_professional
Seguretat i medi ambient                   308
Serveis socioculturals i a la comunitat    264
Formació complementària                    185
Competències transversals                  162
Informàtica i comunicacions                127
Comerç i màrqueting                        121
Administració i gestió                     106
Hostaleria i turisme                       104
Sanitat                                     78
Transport i manteniment de vehicles         77
Edificació i obra civil                     77
Agrària                                     56
Activitats físiques i esportives            51
Fabricació mecànica                         30
Indústries alimentàries                     28
Imatge personal                             21
Electricitat i electrònica                  21
Energia i aigua                             20
Química                                 

In [196]:
import re

# Función para asignar tipo de formación
def asignar_tipo_formacion(row):
    if pd.isna(row['tipus_de_formacio']):
        # Si tiene código normalizado (XXXX####)
        if bool(re.match(r'^[A-Z]{4}\d{4}', row['nom_de_laccio'])):
            return 'Certificats de Professionalitat'
            
        # Por familia profesional
        if row['familia_professional'] in ['Seguretat i medi ambient', 'Formació complementària', 'Competències transversals']:
            return 'Altra Formació Professionalitzadora'
            
        # Por área profesional
        if row['area_professional'] in ['Digitals / Ofimàtica / Tics', 'Formació i educació']:
            return 'Altra Formació Professionalitzadora'
            
        return 'Altra Formació Professionalitzadora'  # Categoría por defecto
        
    return row['tipus_de_formacio']

# Aplicar la función
df['tipus_de_formacio'] = df.apply(asignar_tipo_formacion, axis=1)

# Verificar el resultado
print("Distribución final de tipus_de_formacio:")
print(df['tipus_de_formacio'].value_counts())
print(f"\nValores nulos restantes: {df['tipus_de_formacio'].isna().sum()}")

Distribución final de tipus_de_formacio:
tipus_de_formacio
Altra Formació Professionalitzadora                                                                                                          3784
Certificats de Professionalitat                                                                                                               866
Acreditacions oficials que capaciten per a un lloc de treball, formació regulada per conveni col·lectiu o certificacions de fabricant TIC     172
Formació d'idiomes preparatòria per certificacions segons Marc Comú Europeu de Referència de les Llengües                                     130
Name: count, dtype: int64

Valores nulos restantes: 0


- Anàlisis de la variable codi_comarca / nom_comarca y codi_municipi / nom_municipi

In [197]:
# Análisis de variables geográficas
print("1. Análisis comarca:")
print("\nValores únicos codi_comarca vs nom_comarca:")
print(df[['codi_comarca', 'nom_comarca']].drop_duplicates().sort_values('codi_comarca'))

print("\n2. Análisis municipi:")
print("\nValores únicos codi_municipi vs nom_municipi:")
print(df[['codi_municipi', 'nom_municipi']].drop_duplicates().sort_values('codi_municipi'))

# Verificar si hay nulos
print("\n3. Valores nulos:")
print(df[['codi_comarca', 'nom_comarca', 'codi_municipi', 'nom_municipi']].isna().sum())

1. Análisis comarca:

Valores únicos codi_comarca vs nom_comarca:
               codi_comarca          nom_comarca
IDENTIFICADOR                                   
260037                    0  NORTE METROPOLITANO
260476                    0  OESTE METROPOLITANO
260327                    3          ALT PENEDÈS
258172                    7                BAGES
253904                    8            BAIX CAMP
259557                    9            BAIX EBRE
257005                   10         BAIX EMPORDÀ
259784                   11       BAIX LLOBREGAT
251433                   13           BARCELONÈS
251302                   19             GARROTXA
258439                   20              GIRONÈS
244863                   21              MARESME
253285                   22              MONTSIÀ
239981                   24                OSONA
257176                   30        RIBERA D'EBRE
258893                   31             RIPOLLÈS
258262                   33               SEGRIÀ
260

In [198]:
# 1. Verificar la correspondencia entre códigos y nombres
print("1. Verificación de correspondencia única:")
comarca_check = df.groupby('codi_comarca')['nom_comarca'].nunique()
municipi_check = df.groupby('codi_municipi')['nom_municipi'].nunique()

print("\nComarcas con más de un nombre:")
print(comarca_check[comarca_check > 1])

print("\nMunicipios con más de un nombre:")
print(municipi_check[municipi_check > 1])

# 2. Distribución de registros por comarca
print("\n2. Top 10 comarcas por número de registros:")
print(df.groupby(['codi_comarca', 'nom_comarca']).size().sort_values(ascending=False).head(10))

# 3. Verificar consistencia de códigos
print("\n3. Verificar si hay códigos sin nombre o nombres sin código:")
inconsistencias = df[
    (df['codi_comarca'].notna() & df['nom_comarca'].isna()) |
    (df['codi_comarca'].isna() & df['nom_comarca'].notna()) |
    (df['codi_municipi'].notna() & df['nom_municipi'].isna()) |
    (df['codi_municipi'].isna() & df['nom_municipi'].notna())
]
print(f"Registros con inconsistencias: {len(inconsistencias)}")

1. Verificación de correspondencia única:

Comarcas con más de un nombre:
codi_comarca
0    2
Name: nom_comarca, dtype: int64

Municipios con más de un nombre:
codi_municipi
8266    2
Name: nom_municipi, dtype: int64

2. Top 10 comarcas por número de registros:
codi_comarca  nom_comarca        
13            BARCELONÈS             3246
40            VALLÈS OCCIDENTAL       473
41            VALLÈS ORIENTAL         180
11            BAIX LLOBREGAT          177
33            SEGRIÀ                  122
20            GIRONÈS                 122
8             BAIX CAMP               106
30            RIBERA D'EBRE           100
0             OESTE METROPOLITANO      75
31            RIPOLLÈS                 66
dtype: int64

3. Verificar si hay códigos sin nombre o nombres sin código:
Registros con inconsistencias: 0


In [199]:
# Ver detalle de la inconsistencia en comarca
print("Detalle de la comarca con código 0:")
print(df[df['codi_comarca'] == 0][['codi_comarca', 'nom_comarca']].drop_duplicates())

Detalle de la comarca con código 0:
               codi_comarca          nom_comarca
IDENTIFICADOR                                   
260476                    0  OESTE METROPOLITANO
260037                    0  NORTE METROPOLITANO


In [200]:
# Ver detalle de la inconsistencia en municipio
print("\nDetalle del municipio con código 8266:")
print(df[df['codi_municipi'] == 8266][['codi_municipi', 'nom_municipi']].drop_duplicates())


Detalle del municipio con código 8266:
               codi_municipi           nom_municipi
IDENTIFICADOR                                      
256855                  8266  CERDANYOLA DEL VALLÈS
259088                  8266  CERDANYOLA DEL VALLES


In [201]:
# 1. Corregir comarca con código 0
def corregir_comarca(row):
    if row['codi_comarca'] == 0:
        if row['nom_comarca'] == "OESTE METROPOLITANO":
            return "ÀREA METROPOLITANA OEST"
        elif row['nom_comarca'] == "NORTE METROPOLITANO":
            return "ÀREA METROPOLITANA NORD"
    return row['nom_comarca']

# 2. Corregir municipio 8266 (normalizar a catalán con acento)
def corregir_municipi(row):
    if row['codi_municipi'] == 8266:
        return "CERDANYOLA DEL VALLÈS"
    return row['nom_municipi']

# Aplicar correcciones
df['nom_comarca'] = df.apply(corregir_comarca, axis=1)
df['nom_municipi'] = df.apply(corregir_municipi, axis=1)

# Verificar correcciones
print("Verificación después de correcciones:")
print("\n1. Comarcas con código 0:")
print(df[df['codi_comarca'] == 0][['codi_comarca', 'nom_comarca']].drop_duplicates())

print("\n2. Municipio 8266:")
print(df[df['codi_municipi'] == 8266][['codi_municipi', 'nom_municipi']].drop_duplicates())

# Verificar que no quedan duplicados
print("\n3. Verificación final de unicidad:")
print("Comarcas con más de un nombre:")
print(df.groupby('codi_comarca')['nom_comarca'].nunique()[df.groupby('codi_comarca')['nom_comarca'].nunique() > 1])
print("\nMunicipios con más de un nombre:")
print(df.groupby('codi_municipi')['nom_municipi'].nunique()[df.groupby('codi_municipi')['nom_municipi'].nunique() > 1])

Verificación después de correcciones:

1. Comarcas con código 0:
               codi_comarca              nom_comarca
IDENTIFICADOR                                       
260476                    0  ÀREA METROPOLITANA OEST
260037                    0  ÀREA METROPOLITANA NORD

2. Municipio 8266:
               codi_municipi           nom_municipi
IDENTIFICADOR                                      
256855                  8266  CERDANYOLA DEL VALLÈS

3. Verificación final de unicidad:
Comarcas con más de un nombre:
codi_comarca
0    2
Name: nom_comarca, dtype: int64

Municipios con más de un nombre:
Series([], Name: nom_municipi, dtype: int64)


- Anàlisis de la variable familia_profesional

In [202]:
# 1. Análisis de familia_professional y area_professional
print("1. Distribución de familia_professional:")
print(df['familia_professional'].value_counts())

print("\n2. Relación entre familia_professional y area_professional:")
relacion = df.groupby('familia_professional')['area_professional'].nunique()
print("\nNúmero de áreas profesionales por familia:")
print(relacion.sort_values(ascending=False))

print("\n3. Ejemplo de distribución para una familia específica:")
familia_ejemplo = df['familia_professional'].value_counts().index[0]
print(f"\nÁreas profesionales en {familia_ejemplo}:")
print(df[df['familia_professional'] == familia_ejemplo]['area_professional'].value_counts())

# 4. Verificar valores nulos
print("\n4. Valores nulos:")
print(df[['familia_professional', 'area_professional']].isna().sum())

1. Distribución de familia_professional:
familia_professional
Serveis socioculturals i a la comunitat    978
Competències transversals                  699
Administració i gestió                     664
Comerç i màrqueting                        481
Seguretat i medi ambient                   413
Hostaleria i turisme                       266
Formació complementària                    212
Sanitat                                    197
Informàtica i comunicacions                194
Edificació i obra civil                    136
Indústries alimentàries                    108
Agrària                                    102
Transport i manteniment de vehicles         92
Fabricació mecànica                         88
Activitats físiques i esportives            55
Química                                     44
Electricitat i electrònica                  44
Arts gràfiques                              38
Instal·lació i manteniment                  35
Imatge personal                             3

- Anàlisis de la variable hores

In [203]:
# 1. Análisis estadístico básico
print("1. Estadísticas descriptivas de hores:")
print(df['hores'].describe())

# 2. Distribución de valores
print("\n2. Valores más frecuentes:")
print(df['hores'].value_counts().head(10))

# 3. Análisis por familia profesional
print("\n3. Promedio de horas por familia profesional (top 10):")
print(df.groupby('familia_professional')['hores'].mean().sort_values(ascending=False).head(10))

# 4. Verificar valores atípicos
Q1 = df['hores'].quantile(0.25)
Q3 = df['hores'].quantile(0.75)
IQR = Q3 - Q1
outliers = df[(df['hores'] < (Q1 - 1.5 * IQR)) | (df['hores'] > (Q3 + 1.5 * IQR))]

print("\n4. Valores atípicos:")
print(f"Número de outliers: {len(outliers)}")
print("\nEjemplos de cursos con valores atípicos:")
print(outliers[['familia_professional', 'nom_de_laccio', 'hores']].head())

# 5. Verificar valores nulos o cero
print("\n5. Verificación de valores problemáticos:")
print(f"Valores nulos: {df['hores'].isna().sum()}")
print(f"Valores cero: {len(df[df['hores'] == 0])}")

1. Estadísticas descriptivas de hores:
count    4952.000000
mean       44.114499
std        41.276492
min         4.000000
25%        20.000000
50%        30.000000
75%        60.000000
max       800.000000
Name: hores, dtype: float64

2. Valores más frecuentes:
hores
30    1032
20     711
60     567
40     410
50     211
15     172
10     161
8      140
12     135
80     126
Name: count, dtype: int64

3. Promedio de horas por familia profesional (top 10):
familia_professional
Activitats físiques i esportives           82.000000
Informàtica i comunicacions                71.195876
Imatge personal                            61.677419
Transport i manteniment de vehicles        61.467391
Química                                    58.068182
Serveis socioculturals i a la comunitat    56.391616
Electricitat i electrònica                 52.727273
Fabricació mecànica                        52.397727
Arts gràfiques                             52.236842
Instal·lació i manteniment               

In [204]:
# 1. Crear categorías de duración para análisis
df['categoria_duracion'] = pd.cut(df['hores'], 
                                bins=[0, 20, 40, 80, 800],
                                labels=['Corta', 'Media', 'Larga', 'Extensa'])

# 2. Verificar valores extremos (>200 horas)
cursos_largos = df[df['hores'] > 200]
print("Cursos con más de 200 horas:")
print(cursos_largos[['nom_de_laccio', 'hores', 'familia_professional']].head())

Cursos con más de 200 horas:
                                                   nom_de_laccio  hores  \
IDENTIFICADOR                                                             
259311         SSCS0108_MF0250_Atenció i suport psicosocial d...    210   
260309         SEAG0210_MF0073_Funcionament i operació dels p...    220   
259774         SSCS0108_MF0250_Atenció i suport psicosocial d...    210   
259439         ADGD0208_MF0237_Gestió administrativa de les r...    210   
257765         SSCS0108_MF0250_Atenció i suport psicosocial d...    210   

                                  familia_professional  
IDENTIFICADOR                                           
259311         Serveis socioculturals i a la comunitat  
260309                        Seguretat i medi ambient  
259774         Serveis socioculturals i a la comunitat  
259439                          Administració i gestió  
257765         Serveis socioculturals i a la comunitat  


- Anàlisis variables de contacte

In [205]:
# 1. Verificar valores nulos
print("1. Valores nulos en variables de contacto:")
print(df[['telefon', 'correu_electronic', 'web']].isna().sum())

# 2. Verificar formato de teléfono
print("\n2. Ejemplos de formatos de teléfono:")
print(df['telefon'].value_counts().head())

# 3. Verificar formato de correo
print("\n3. Ejemplos de formatos de correo:")
print(df['correu_electronic'].value_counts().head())

# 4. Verificar formato de web
print("\n4. Ejemplos de formatos de web:")
print(df['web'].value_counts().head())

# 5. Verificar consistencia entre entidades y contacto
print("\n5. Número de contactos por entidad:")
contactos_por_entidad = df.groupby('entitat')[['telefon', 'correu_electronic', 'web']].nunique()
print(contactos_por_entidad.describe())

1. Valores nulos en variables de contacto:
telefon              14
correu_electronic    14
web                  14
dtype: int64

2. Ejemplos de formatos de teléfono:
telefon
934964500.0    486
932172515.0    311
932689292.0    286
934812717.0    212
935210485.0    206
Name: count, dtype: int64

3. Ejemplos de formatos de correo:
correu_electronic
formacio@pimec.org                        486
info@catformacio.org                      311
sroman@idfo.com                           286
fppuerto@ccoo.cat                         212
formaciocatalunya@fundacionlaboral.org    206
Name: count, dtype: int64

4. Ejemplos de formatos de web:
web
www.pimecformacio.cat                                             486
www.catformacio.com                                               311
www.idfo.com                                                      286
fundaciopacopuerto.cat                                            212
https://www.cursosenconstruccion.com/cursos-gratuitos/cataluna    206
Name: co

In [206]:
# 1. Convertir teléfono a formato string y limpiar
df['telefon'] = df['telefon'].fillna('').astype(str).apply(lambda x: x.replace('.0', '').strip())

# 2. Normalizar formato de webs y limpiar
def normalizar_web(url):
    if pd.isna(url) or str(url).strip() == '':
        return ''
    if not str(url).strip().startswith(('http://', 'https://')):
        return 'https://' + str(url).strip()
    return str(url).strip()

df['web'] = df['web'].apply(normalizar_web)

# 3. Limpiar correos electrónicos
df['correu_electronic'] = df['correu_electronic'].fillna('').astype(str).str.strip()

# 4. Verificar registros sin contacto (considerando strings vacíos)
registros_sin_contacto = df[
    (df['telefon'].str.len() == 0) & 
    (df['correu_electronic'].str.len() == 0) & 
    (df['web'].str.len() == 0)
]

print("Registros sin información de contacto (incluyendo strings vacíos):")
print(registros_sin_contacto[['entitat', 'familia_professional']].head())

# Mostrar cantidad total de registros sin contacto
print(f"\nTotal de registros sin información de contacto: {len(registros_sin_contacto)}")

Registros sin información de contacto (incluyendo strings vacíos):
              entitat              familia_professional
IDENTIFICADOR                                          
259734          ETEVA          Seguretat i medi ambient
259730          ETEVA  Activitats físiques i esportives
259737          ETEVA          Seguretat i medi ambient
259736          ETEVA          Seguretat i medi ambient
265581          ETEVA  Activitats físiques i esportives

Total de registros sin información de contacto: 14


- Anàlisis de la variable d'entitat nif_entitat_formacio, entitat

In [207]:
# 1. Verificar la relación entre NIF y entidad
print("1. Verificar correspondencia NIF-Entidad:")
relacion_nif_entidad = df.groupby('nif_entitat_formacio')['entitat'].nunique()
print("\nNIFs con más de una entidad:")
print(relacion_nif_entidad[relacion_nif_entidad > 1])

print("\n2. Distribución de cursos por entidad (top 10):")
print(df['entitat'].value_counts().head(10))

# 3. Verificar formato de NIF
print("\n3. Ejemplos de formato NIF:")
print(df['nif_entitat_formacio'].value_counts().head())

# 4. Verificar valores nulos o problemáticos
print("\n4. Valores nulos:")
print(df[['nif_entitat_formacio', 'entitat']].isna().sum())

1. Verificar correspondencia NIF-Entidad:

NIFs con más de una entidad:
Series([], Name: entitat, dtype: int64)

2. Distribución de cursos por entidad (top 10):
entitat
PIME Coneixement, S.L.                               486
Catformació                                          311
IDFO                                                 286
FUNDACIÓ PER A LA FORMACIÓ I L'ESTUDI PACO PUERTO    212
FUNDACION LABORAL DE LA CONSTRUCCION                 206
FOMENT FORMACIÓ                                      160
Federació de Cooperatives Agràries de Catalunya      144
Formar-te                                            139
UPM                                                  138
BRM PREVIFORM, S.L.                                  100
Name: count, dtype: int64

3. Ejemplos de formato NIF:
nif_entitat_formacio
B62765201    486
G63127708    311
G60189057    286
G61676128    212
G80468416    206
Name: count, dtype: int64

4. Valores nulos:
nif_entitat_formacio    0
entitat                 0
dty

- Aàlisis variable nom_de_laccio

In [208]:
# 1. Análisis básico
print("1. Estadísticas básicas:")
print(f"Total de cursos únicos: {df['nom_de_laccio'].nunique()}")
print(f"Valores nulos: {df['nom_de_laccio'].isna().sum()}")

# 2. Patrones en nombres de cursos
print("\n2. Ejemplos de nombres de cursos más frecuentes:")
print(df['nom_de_laccio'].value_counts().head(10))

# 3. Análisis de formato
def analizar_formato_curso(nombre):
    if pd.isna(nombre):
        return 'nulo'
    if bool(re.match(r'^[A-Z]{4}\d{4}', nombre)):
        return 'certificado'
    return 'otro'

df['tipo_nombre'] = df['nom_de_laccio'].apply(analizar_formato_curso)
print("\n3. Distribución por tipo de formato:")
print(df['tipo_nombre'].value_counts())

# 4. Longitud de nombres
df['longitud_nombre'] = df['nom_de_laccio'].str.len()
print("\n4. Estadísticas de longitud de nombres:")
print(df['longitud_nombre'].describe())

# 5. Relación con familia profesional
print("\n5. Número promedio de cursos por familia profesional:")
print(df.groupby('familia_professional')['nom_de_laccio'].nunique().sort_values(ascending=False).head())

1. Estadísticas básicas:
Total de cursos únicos: 1344
Valores nulos: 0

2. Ejemplos de nombres de cursos más frecuentes:
nom_de_laccio
Comptabilitat bàsica                        58
Nòmines i seguretat social I                45
Comptabilitat avançada                      45
Nòmines i seguretat social II               43
Mindfulness                                 43
Resolució de conflictes                     42
Operació de carretons Norma UNE 58451       36
Gestió de les xarxes socials a l'empresa    33
Habilitats de comunicació                   32
Eines de coaching. Nivell I                 31
Name: count, dtype: int64

3. Distribución por tipo de formato:
tipo_nombre
otro           4086
certificado     866
Name: count, dtype: int64

4. Estadísticas de longitud de nombres:
count    4952.000000
mean       48.237278
std        28.558006
min         7.000000
25%        27.000000
50%        41.000000
75%        64.000000
max       205.000000
Name: longitud_nombre, dtype: float64

5. Nú

In [209]:
# 1. Normalizar nombres de cursos
def normalizar_nombre_curso(nombre):
    # Eliminar espacios múltiples
    nombre = ' '.join(nombre.split())
    # Normalizar mayúsculas/minúsculas
    if not nombre.startswith(('ADGG', 'SSCS', 'SEAG')):  # Preservar códigos de certificados
        nombre = nombre.capitalize()
    return nombre

df['nom_de_laccio'] = df['nom_de_laccio'].apply(normalizar_nombre_curso)

# 2. Crear categoría de nivel
df['nivel_curso'] = df['nom_de_laccio'].str.extract(r'(I{1,3}|IV|V|VI)$')

# 3. Identificar certificados profesionales
df['es_certificado'] = df['nom_de_laccio'].str.match(r'^[A-Z]{4}\d{4}')

In [210]:
# 1. Eliminar columna nivel_curso que está completamente vacía
if 'nivel_curso' in df.columns:
    df = df.drop('nivel_curso', axis=1)

# 2. Mejorar la detección de niveles
def detectar_nivel_mejorado(row):
    nombre = row['nom_de_laccio'].lower()
    
    # Patrones de nivel
    if any(x in nombre for x in ['inicial', 'introducció', 'bàsic', 'basic']):
        return 'Bàsic'
    elif any(x in nombre for x in ['mitjà', 'intermedi', 'mig']):
        return 'Mitjà'
    elif any(x in nombre for x in ['avançat', 'superior', 'especialitzat', 'avancat']):
        return 'Avançat'
    elif nombre.strip().endswith(' i'):
        return 'Nivell I'
    elif nombre.strip().endswith(' ii'):
        return 'Nivell II'
    elif nombre.strip().endswith(' iii'):
        return 'Nivell III'
    
    # Para certificados profesionales
    if bool(re.match(r'^[A-Z]{4}\d{4}', row['nom_de_laccio'])):
        if '_BASIC_' in nombre:
            return 'Bàsic'
        elif '_MIG_' in nombre:
            return 'Mitjà'
        elif '_SUPERIOR_' in nombre:
            return 'Avançat'
    
    return None

# Aplicar la función mejorada
df['nivell'] = df.apply(detectar_nivel_mejorado, axis=1)

# Mostrar resultados
print("1. Nueva distribución de niveles:")
print(df['nivell'].value_counts(dropna=False))

print("\n2. Ejemplos de cursos con nivel detectado:")
print(df[df['nivell'].notna()][['nom_de_laccio', 'familia_professional', 'nivell']].head(10))

# 3. Verificar consistencia
print("\n3. Distribución de niveles por familia profesional:")
print(pd.crosstab(df['familia_professional'], df['nivell'], margins=True))

1. Nueva distribución de niveles:
nivell
None          4183
Bàsic          355
Nivell I       150
Nivell II      110
Avançat         96
Mitjà           56
Nivell III       2
Name: count, dtype: int64

2. Ejemplos de cursos con nivel detectado:
                                                   nom_de_laccio  \
IDENTIFICADOR                                                      
241674         Manipulació i aplicació de productes fitosanit...   
258446                Comt0211_mf1329_atenció bàsica als clients   
250718         Resolució de problemes a l'entorn digital (act...   
241673         Manipulació i aplicació de productes fitosanit...   
259073         ADGG0508_UF0511_Tractament bàsic de dades i fu...   
251344           Principis bàsics de la intel·ligència emocional   
259565         Ssce0110_mf1443_selecció, elaboració, adaptaci...   
258457         Ssce0110_mf1443_selecció, elaboració, adaptaci...   
249107                                      Comptabilitat bàsica   
255822  

- Corrección

In [211]:
# # Eliminar columnas con un único valor
# columnas_vacias = [col for col in df.columns if df[col].nunique() == 1]
# df = df.drop(columns=columnas_vacias)
# print("Columnas eliminadas:", columnas_vacias)

### DUPLICADOS

- Identificación

In [212]:
df.duplicated().sum()

np.int64(131)

In [213]:
# 1. Verificar duplicados exactos
print("1. Análisis de duplicados:")
print(f"Total de filas duplicadas: {df.duplicated().sum()}")

# 2. Ver qué columnas están causando los duplicados
print("\n2. Duplicados por combinaciones de columnas:")
columnas_clave = ['nom_de_laccio', 'entitat', 'familia_professional', 'area_professional', 'modalitat', 'hores']
duplicados = df[df.duplicated(subset=columnas_clave, keep=False)].sort_values(by=columnas_clave)
print(f"\nDuplicados considerando columnas clave: {len(duplicados)}")

# 3. Mostrar ejemplos de duplicados
print("\n3. Ejemplos de registros duplicados:")
print(duplicados[columnas_clave].head(10))

# 4. Análisis de duplicados por entidad
print("\n4. Número de duplicados por entidad:")
print(duplicados['entitat'].value_counts().head())

1. Análisis de duplicados:
Total de filas duplicadas: 131

2. Duplicados por combinaciones de columnas:

Duplicados considerando columnas clave: 262

3. Ejemplos de registros duplicados:
                                                   nom_de_laccio  \
IDENTIFICADOR                                                      
243394         Actualització plataformes elevadores mòbils de...   
243061         Actualització plataformes elevadores mòbils de...   
253903         Actualització plataformes elevadores mòbils de...   
253713         Actualització plataformes elevadores mòbils de...   
249435         Actualització plataformes elevadores mòbils de...   
249690         Actualització plataformes elevadores mòbils de...   
250478                            Analítica i màrqueting digital   
250667                            Analítica i màrqueting digital   
249399                            Analítica i màrqueting digital   
249692                            Analítica i màrqueting digital 

- Corrección

In [214]:
# 1. Crear identificador único de curso
df['id_curso'] = df.apply(lambda x: f"{x['nom_de_laccio']}_{x['entitat']}_{x['modalitat']}_{x['hores']}", axis=1)

# 2. Identificar duplicados manteniendo el primer registro
duplicados = df[df.duplicated(['id_curso'], keep='first')]

# 3. Eliminar duplicados manteniendo el primer registro
df_sin_duplicados = df.drop_duplicates(['id_curso'], keep='first')

# Verificar resultados
print("1. Estadísticas después de eliminar duplicados:")
print(f"Registros originales: {len(df)}")
print(f"Registros únicos: {len(df_sin_duplicados)}")
print(f"Duplicados eliminados: {len(df) - len(df_sin_duplicados)}")

# 4. Guardar registro de duplicados eliminados
print("\n2. Ejemplos de duplicados eliminados:")
print(duplicados[['nom_de_laccio', 'entitat', 'modalitat', 'hores']].head())

1. Estadísticas después de eliminar duplicados:
Registros originales: 4952
Registros únicos: 4821
Duplicados eliminados: 131

2. Ejemplos de duplicados eliminados:
                                                   nom_de_laccio  \
IDENTIFICADOR                                                      
259392         Ssce0110_mp0353_mòdul de pràctiques profession...   
242826                         Elaboració d'un pla de màrqueting   
253714         Operador/a de plataformes mòbils de personal (...   
251336                                    Comptabilitat avançada   
249996                              Comptabilitat informatitzada   

                           entitat     modalitat  hores  
IDENTIFICADOR                                            
259392                       INKOR    Presencial     40  
242826                         UPM    Presencial     60  
253714         MOTIVACCIO FORMACIO    Presencial     12  
251336              ADAMS FORMACIÓ  Teleformació     60  
249996     

### SEPARAR NUMÉRICAS Y CATEGÓRICAS

**Categóricas**

In [215]:
cat = df.select_dtypes(exclude = 'number').copy()

**Numéricas**

In [216]:
num = df.select_dtypes(include='number').copy()

### GESTIÓN DE CATEGÓRICAS

#### NULOS

Identificación

In [217]:
cat.isna().sum().sort_values(ascending = False)

nivell                  4183
familia_professional       0
area_professional          0
nif_entitat_formacio       0
nom_de_laccio              0
nom_municipi               0
nom_comarca                0
modalitat                  0
entitat                    0
tipus_de_formacio          0
web                        0
telefon                    0
correu_electronic          0
categoria_duracion         0
tipo_nombre                0
es_certificado             0
id_curso                   0
dtype: int64

In [218]:
def detectar_nivel_v5(row):
    nombre = row['nom_de_laccio'].lower()
    familia = row['familia_professional'].lower()
    
    # 1. Patrones específicos por familia profesional
    patrones_especificos = {
        'informàtica i comunicacions': {
            'Bàsic': ['usuari', 'iniciació', 'fonaments', 'introducció', 'bàsic'],
            'Mitjà': ['administració', 'desenvolupament', 'programació', 'tècnic'],
            'Avançat': ['arquitectura', 'seguretat', 'cloud', 'expert', 'especialista']
        },
        'administració i gestió': {
            'Bàsic': ['auxiliar', 'operacions', 'bàsic', 'inicial'],
            'Mitjà': ['gestió', 'administratiu', 'comptable'],
            'Avançat': ['direcció', 'estratègia', 'superior']
        },
        'idiomes': {
            'Bàsic': ['a1', 'a2', 'basic', 'inicial'],
            'Mitjà': ['b1', 'b2', 'intermediate'],
            'Avançat': ['c1', 'c2', 'advanced']
        }
    }
    
    # 2. Patrones de idiomas específicos
    match_idioma = re.search(r'([abcd][12]?\.[12]?)', nombre)
    if match_idioma:
        nivel_idioma = match_idioma.group(1)
        return f'Idiomes {nivel_idioma.upper()}'
    
    # 3. Módulos de prácticas
    if any(x in nombre for x in ['pràctiques professionals', 'mp0', '_mp_']):
        return 'Pràctiques'
    
    # 4. Verificar patrones específicos por familia
    for fam, patrones in patrones_especificos.items():
        if fam in familia:
            for nivel, palabras in patrones.items():
                if any(p in nombre for p in palabras):
                    return nivel
    
    # 5. Certificados profesionales
    if bool(re.match(r'^[A-Z]{4}\d{4}', row['nom_de_laccio'])):
        if any(x in nombre for x in ['_basic_', '_inicial_', '_uf0']):
            return 'Bàsic'
        elif any(x in nombre for x in ['_mig_', '_intermedi_']):
            return 'Mitjà'
        elif any(x in nombre for x in ['_superior_', '_avancat_']):
            return 'Avançat'
    
    # 6. Patrones generales
    patrones_generales = {
        'Bàsic': ['inicial', 'introducció', 'bàsic', 'basic', 'iniciació', 
                 'elemental', 'principis', 'fonaments', 'primer nivell'],
        'Mitjà': ['mitjà', 'intermedi', 'mig', 'segon nivell', 
                 'perfeccionament', 'especialització'],
        'Avançat': ['avançat', 'superior', 'especialitzat', 'expert', 
                   'master', 'tercer nivell']
    }
    
    for nivel, patrones in patrones_generales.items():
        if any(p in nombre for p in patrones):
            return nivel
    
    # 7. Números romanos
    if re.search(r'\bI\b$', nombre):
        return 'Nivell I'
    elif re.search(r'\bII\b$', nombre):
        return 'Nivell II'
    elif re.search(r'\bIII\b$', nombre):
        return 'Nivell III'
    
    return None

# Aplicar la función mejorada
df['nivell'] = df.apply(detectar_nivel_v5, axis=1)

# Análisis de resultados
print("1. Distribución general de niveles:")
print(df['nivell'].value_counts(dropna=False))

print("\n2. Distribución por familia profesional:")
print(pd.crosstab(df['familia_professional'], df['nivell']).head())

print("\n3. Ejemplos de detección por nivel:")
for nivel in df['nivell'].unique():
    if nivel is not None:
        print(f"\nNivel: {nivel}")
        print(df[df['nivell'] == nivel][['nom_de_laccio', 'familia_professional']].head(2))

1. Distribución general de niveles:
nivell
None           3639
Bàsic           455
Mitjà           201
Pràctiques      150
Avançat         142
Idiomes B.2      60
Idiomes B.1      59
Idiomes A.1      58
Idiomes A.2      43
Idiomes C.1      40
Idiomes C.2      39
Idiomes A.       30
Idiomes C.       30
Idiomes D.        6
Name: count, dtype: int64

2. Distribución por familia profesional:
nivell                            Avançat  Bàsic  Idiomes A.  Idiomes A.1  \
familia_professional                                                        
Activitats físiques i esportives        0      0           0            0   
Administració i gestió                 13    110           0            0   
Agrària                                 7     13           0            0   
Arts gràfiques                          0      0           0            0   
Arts i artesanies                       0      0           2            0   

nivell                            Idiomes A.2  Idiomes B.1  Idiomes B

In [219]:
# 1. Consolidar niveles similares
df['nivel_consolidado'] = df['nivell'].map(lambda x: 
    'Bàsic' if x in ['Bàsic', 'Idiomes A.', 'Idiomes A.1', 'Idiomes A.2'] else
    'Mitjà' if x in ['Mitjà', 'Idiomes B.1', 'Idiomes B.2'] else
    'Avançat' if x in ['Avançat', 'Idiomes C.', 'Idiomes C.1', 'Idiomes C.2', 'Idiomes D.'] else
    'Pràctiques' if x == 'Pràctiques' else
    'Sense nivell'
)

# 2. Crear columna de tipo de nivel
df['tipus_nivell'] = df['nivell'].map(lambda x:
    'Idiomes' if str(x).startswith('Idiomes') else
    'Pràctiques' if x == 'Pràctiques' else
    'General' if x in ['Bàsic', 'Mitjà', 'Avançat'] else
    'Sense classificar'
)

# Mostrar resultados consolidados
print("1. Distribución de niveles consolidados:")
print(df['nivel_consolidado'].value_counts())

print("\n2. Distribución por tipo de nivel:")
print(df['tipus_nivell'].value_counts())

1. Distribución de niveles consolidados:
nivel_consolidado
Sense nivell    3639
Bàsic            586
Mitjà            320
Avançat          257
Pràctiques       150
Name: count, dtype: int64

2. Distribución por tipo de nivel:
tipus_nivell
Sense classificar    3639
General               798
Idiomes               365
Pràctiques            150
Name: count, dtype: int64


In [220]:
# 1. Crear categorización completa
df['categoritzacio_curs'] = df.apply(lambda row: {
    'nivell': row['nivel_consolidado'],
    'tipus': row['tipus_nivell'],
    'familia': row['familia_professional'],
    'modalitat': row['modalitat'],
    'hores': row['categoria_duracion']
}, axis=1)

# 2. Análisis cruzado
print("1. Distribución de niveles por modalidad:")
print(pd.crosstab(df['nivel_consolidado'], df['modalitat']))

print("\n2. Distribución de tipos por familia profesional:")
print(pd.crosstab(df['familia_professional'], df['tipus_nivell']))

# 3. Estadísticas por nivel
for nivel in df['nivel_consolidado'].unique():
    if nivel != 'Sense nivell':
        subset = df[df['nivel_consolidado'] == nivel]
        print(f"\nEstadísticas para nivel {nivel}:")
        print(f"Total cursos: {len(subset)}")
        print(f"Familias principales:")
        print(subset['familia_professional'].value_counts().head(3))
        print(f"Duración media: {subset['hores'].mean():.1f} horas")

1. Distribución de niveles por modalidad:
modalitat          Mixta  Presencial  Teleformació
nivel_consolidado                                 
Avançat                1         200            56
Bàsic                  6         467           113
Mitjà                  1         228            91
Pràctiques             0         150             0
Sense nivell           7        2776           856

2. Distribución de tipos por familia profesional:
tipus_nivell                             General  Idiomes  Pràctiques  \
familia_professional                                                    
Activitats físiques i esportives               0        0          10   
Administració i gestió                       250        0          16   
Agrària                                       20        0           0   
Arts gràfiques                                 0        0           1   
Arts i artesanies                              0        2           0   
Comerç i màrqueting                    

1. La modalidad presencial domina (77.2%)
2. Las prácticas son exclusivamente presenciales
3. Los cursos básicos tienen mayor presencia en todas las modalidades
4. La duración media varía significativamente por nivel
5. Servicios socioculturales y Administración son las familias más diversas

Corrección

- Eliminar la variable

In [221]:
# Opción 1: Eliminar estas columnas por tener demasiados nulos
# cat = cat.drop(columns=['notes_equipament', 'notes_dades'])

# # Opción 2: Mantenerlas pero imputar con un valor específico
# cat['notes_equipament'] = cat['notes_equipament'].fillna('Sin notas')
# cat['notes_dades'] = cat['notes_dades'].fillna('Sin notas')

# Opción 3: Mantenerlas tal cual si la información, aunque escasa, es valiosa

- Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

In [222]:
# criterio = 3
# cat.dropna(thresh=criterio, inplace=True)
# num.dropna(thresh=criterio, inplace=True)

- Imputar por un valor

Variables a imputar

In [223]:
# var_imputar_valor = ['resultado_campana_anterior','formacion']

Valor por el que imputar

In [224]:
# valor = 'NULO'

- Imputar

In [225]:
# cat[var_imputar_valor] = cat[var_imputar_valor].fillna(valor)

- Imputar por la moda

Variables a imputar

In [226]:
# var_imputar_moda = ['resultado_campana_anterior','formacion']

Funcion para imputar con la moda

In [227]:
# def imputar_moda(variable):
#     return(variable.fillna(variable.mode()[0]))

Imputar con la moda

In [228]:
# cat[var_imputar_moda] = cat[var_imputar_moda].apply(imputar_moda)

### ATÍPICOS

- Categorías poco frecuentes

Función para agrupar categorías raras en 'OTROS'

In [229]:
# def agrupar_cat_raras(variable, criterio = 0.05):
#     # Calcula las frecuencias
#     frecuencias = variable.value_counts(normalize=True)
#     # Identifica las que están por debajo del criterio
#     temp = [cada for cada in frecuencias.loc[frecuencias < criterio].index.values]
#     # Las recodifica en 'OTROS'
#     temp2 = np.where(variable.isin(temp),'OTROS',variable)
#     # Devuelve el resultado
#     return(temp2)

Variables a imputar

In [230]:
# var_imputar_moda = ['trabajo','estado_civil']

Aplicar

In [231]:
# for variable in var_imputar_moda:
#     cat[variable] = agrupar_cat_raras(cat[variable],criterio = 0.02)

### GESTIÓN DE NUMÉRICAS

#### NULOS

Identificación

In [232]:
num.isna().sum().sort_values(ascending = False)

codi_municipi      0
codi_comarca       0
hores              0
longitud_nombre    0
dtype: int64

Corrección

- Eliminar la variable

In [233]:
# num.drop(columns = 'num_contactos_otras_campanas',inplace = True)

- Eliminar registros

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

In [234]:
# criterio = 3
# cat.dropna(thresh=criterio, inplace=True)
# num.dropna(thresh=criterio, inplace=True)

- Imputar por la media

Variables a imputar

In [235]:
# var_imputar_media = ['edad','num_contactos_esta_campana']

Funcion para imputar con la media

In [236]:
# def imputar_media(variable):
#     if pd.api.types.is_integer_dtype(variable):
#         return(variable.fillna(int(variable.mean())))
#     else:
#         return(variable.fillna(variable.mean()))

- Imputar con la media

In [237]:
# num[var_imputar_media] = num[var_imputar_media].apply(imputar_media)

- Imputar por la mediana

Variables a imputar

In [238]:
# var_imputar_mediana = ['edad','num_contactos_esta_campana']

Funcion para imputar con la mediana

In [239]:
# def imputar_mediana(variable):
#     if pd.api.types.is_integer_dtype(variable):
#         return(variable.fillna(int(variable.median())))
#     else:
#         return(variable.fillna(variable.median()))

- Imputar con la mediana

In [240]:
# num[var_imputar_mediana] = num[var_imputar_mediana].apply(imputar_mediana)

### ATÍPICOS

**Por desviación típica**

- Identificación

Definir el número de desviaciones típicas

In [241]:
# num_desv_tip = 4

Función que devuelve los índices de los atípicos

In [242]:
# def atipicos_desv_tip(variable, num_desv_tip = 4):
#     #sacamos los nulos por ahora
#     variable = variable.dropna()
#     #calculamos los límites
#     media = np.mean(variable)
#     sd = np.std(variable)
#     umbral = sd * num_desv_tip
#     lim_inf = media - umbral
#     lim_sup = media + umbral
#     #encontramos los índices de los que están fuera de los límites
#     indices = [indice for indice,valor in variable.items() if valor < lim_inf or valor > lim_sup]
#     return(indices)

Función que cuenta el número de atípicos

In [243]:
# def conteo_atipicos(df,variable, num_desv_tip = 4):
#     atipicos = atipicos_desv_tip(df[variable], num_desv_tip)
#     return(df.loc[atipicos,variable].value_counts())

Variables a aplicar

In [244]:
# var_atipicos_dt = ['edad','num_contactos_esta_campana']

Aplicar

In [245]:
# for variable in var_atipicos_dt:
#     print('\n' + variable + ':\n')
#     print(conteo_atipicos(num,variable,num_desv_tip))

- Corrección

Eliminar los registros que tengan al menos un atípico en alguna de las variables indicadas.

*Cuidado: Si se eliminan registros hay que eliminarlos tanto de cat como de num*

Variables a aplicar

In [246]:
# var_atipicos_dt = ['edad','num_contactos_esta_campana']

Aplicar

In [247]:
# for variable in var_atipicos_dt:
#     cat.drop(atipicos_desv_tip(num[variable], num_desv_tip),axis = 0, inplace = True)
#     num.drop(atipicos_desv_tip(num[variable], num_desv_tip),axis = 0, inplace = True)

**Por winsorización manual para una variable** 

- Identificación

Variable a analizar

In [248]:
# variable = 'variacion_tasa_empleo'

Comprobar el mínimo y máximo actual

In [249]:
# num[variable].describe().loc[['min','max']]

- Corrección

Definir el mínimo y máximo tolerable.

In [250]:
# minimo = -20
# maximo = 10

Aplicar

In [251]:
# num[variable].clip(minimo,maximo, inplace=True)

**Por winsorización automática para múltiples variables**

Variables a winsorizar

In [252]:
# var_winsorizar = ['edad','num_contactos_esta_campana']

Definir los percentiles mínimos y máximos.

In [253]:
# p_min = 0.01
# p_max = 0.99

Aplicar

In [254]:
# for variable in var_winsorizar:
#     num[variable].clip(lower = num[variable].quantile(p_min), 
#                        upper = num[variable].quantile(p_max), 
#                        inplace = True)

### GUARDAR DATASETS TRAS CALIDAD DE DATOS

Guarda los avances en df, cat y num con un sufijo para poder guardar o recuperar avances del proyecto.

En formato pickle para no perder las modificaciones de metadatos.

- Definir los nombres de los archivos

In [255]:
ruta_trabajo = ruta_proyecto + '/02_Datos/03_Trabajo/' + '2024_accions_formatives_previstes_consorci.pickle'
ruta_cat = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'cat_2024_accions_formatives_previstes_consorci.pickle'
ruta_num = ruta_proyecto + '/02_Datos/03_Trabajo/' + 'num_2024_accions_formatives_previstes_consorci.pickle'

- Guardar los archivos

In [256]:
df.to_pickle(ruta_trabajo)
cat.to_pickle(ruta_cat)
num.to_pickle(ruta_num)